src/mlpack/core/dists/diag_cov_gaussian_distribution.cpp

```diff
@@ -0,0 +1,170 @@
+{
+  // Calculate Log determinant of the diagonal covariance.
+  logDetCov = arma::accu(log(covariance.diag()));
+}
+
+arma::vec DiagCovGaussianDistribution::Random() const
+{
+  return arma::sqrt(covariance) * arma::randn<arma::vec>(mean.n_elem) + mean;
+}
+
+void DiagCovGaussianDistribution::Train(const arma::mat& observations)
+{
+  if (observations.n_cols > 1)
+  {
+    mean.zeros(observations.n_rows);
+    covariance.zeros(observations.n_rows, observations.n_rows);
+  }
+  else
+  {
+    mean.zeros(0);
+    covariance.zeros(0);
+    return;
+  }
+
+  // Calculate the mean.
+  for (size_t i = 0; i < observations.n_cols; i++)
+    mean += observations.col(i);
+
+  // Normalize the mean.
+  mean /= observations.n_cols;
+
+  // Now calculate the covariance.
+  for (size_t i = 0; i < observations.n_cols; i++)
+  {
+    arma::vec obsNoMean = observations.col(i) - mean;
+    covariance.diag() += arma::diagvec(obsNoMean * trans(obsNoMean));
+  }
+
+  // Finish estimating the covariance by normalizing, with the (1 / (n - 1))
+  // to make the estimator unbiased.
+  covariance.diag() /= (observations.n_cols - 1);
+
+  // Ensure that the covariance is diagonal.
+  gmm::DiagonalConstraint::ApplyConstraint(covariance);
+
+  InverseCovariance();
+  LogDeterminant();
+}
+
+void DiagCovGaussianDistribution::Train(const arma::mat& observations,
+                                        const arma::vec& probabilities)
+{
+  if (observations.n_cols > 0)
+  {
+    mean.zeros(observations.n_rows);
+    covariance.zeros(observations.n_rows, observations.n_rows);
+  }
+  else
+  {
+    mean.zeros(0);
+    covariance.zeros(0);
+    return;
+  }
+
+  // We'll normalize the covariance with (v1 - (v2 / v1))
+  // for unbiased estimator in the weighted arithmetic mean. The v1 is the sum
+  // of the weights, and the v2 is the sum of the each weight squared.
+  // If you want to know more detailed description,
+  // please refer to https://en.wikipedia.org/wiki/Weighted_arithmetic_mean
+  double v1 = arma::accu(probabilities);
+  double v2 = 0;
+  arma::vec normalizedProbs = probabilities;
+
+  // If their sum is 0, there is nothing in this Gaussian.
+  // At least, set the covariance so that it's invertible.
+  if (v1 == 0)
+  {
+    covariance.diag() += 1e-50;
+    InverseCovariance();
+    LogDeterminant();
+    return;
+  }
+
+  // If their sum is not 1, divide the weights by them to normalize.
+  if (v1 != 1)
+  {
+    normalizedProbs = probabilities / v1;
+    v1 = arma::accu(normalizedProbs);
+  }
+
+  // Calculate the mean.
+  for (size_t i = 0; i < observations.n_cols; i++)
+    mean += normalizedProbs[i] * observations.col(i);
+
+  // Now find the covariance.
+  for (size_t i = 0; i < observations.n_cols; i++)
+  {
+    arma::vec obsNoMean = observations.col(i) - mean;
+    covariance.diag() += arma::diagvec(
+        normalizedProbs[i] * (obsNoMean * trans(obsNoMean)));
+    v2 += normalizedProbs(i) * normalizedProbs(i);
+  }
```

Comment:
Actually it may be still faster here to avoid instantiating `onesRow`; Armadillo can optimize the `arma::ones<>` call, so you could do

```
arma::mat obsNoMean = observations - mean * arma::ones<arma::rowvec>(observations.n_cols);
```

Reply:
Aha~, Thanks!